### PR TITLE
Fix handling of globally defined operations in Target

### LINKED ISSUE #8914: no validity check on Target construction
### Environment

- **Qiskit Terra version**: 
- **Python version**: 
- **Operating system**: 

### What is happening?

It's possible to build a Target where the Instruction does not match the number of qubits it's being added to. 

### How can we reproduce the issue?

For example here I add a 1-qubit gate to a two-qubit tuple.

```py
from qiskit.transpiler import Target, InstructionProperties
from qiskit.circuit.library import RZGate
from qiskit.circuit import Parameter

θ = Parameter('θ')
rz_props = {
    (0, 1): InstructionProperties(duration=0, error=0)
}

target = Target()
target.add_instruction(RZGate(θ), rz_props, name='rz')
```

### What should happen?

raise Error 

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes the handling in the target for working with instructions defined globally in the target. It adds handling to the operation_names_for_qargs() and operations_for_qargs() methods to build outputs that include globally defined operations instead of ignoring them. Additionally the add_instruction() method is updated to validate the input qargs to ensure that only valid qubits and qargs are allowed to be added to the target for an instruction.

### Details and comments

Fixes #8914

### CODE REVIEW

File: qiskit/transpiler/target.py [hunk 1]
```diff
@@ -238,6 +239,8 @@ def __init__(
         self._gate_name_map = {}
         # A nested mapping of gate name -> qargs -> properties
         self._gate_map = {}
+        # A mapping of number of qubits to set of op names which are global
+        self._global_operations = defaultdict(set)
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:243]: How much do we care about reducing search times by keying this on `(num_qubits, num_clbits)`?  I suspect the way you've done this is better in the longer term, since qubits _are_ special compared to clbits, but just throwing out the question in case it was an oversight.

File: qiskit/transpiler/target.py [hunk 2]
```diff
@@ -539,30 +550,44 @@ def operations_for_qargs(self, qargs):
...
         """
-        if qargs not in self._qarg_gate_map:
+        if qargs is not None and any(x not in range(0, self.num_qubits) for x in qargs):
             raise KeyError(f"{qargs} not in target.")
         res = [self._gate_name_map[x] for x in self._qarg_gate_map[qargs]]
-        if None in self._qarg_gate_map:
-            res += [self._gate_name_map[x] for x in self._qarg_gate_map[None]]
-        return res
+        if qargs is not None:
+            res += self._global_operations.get(len(qargs), list())
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:557]: I actually thought there was a pylint rule that complained about using `list()` or `dict()` in place of the empty literals, but if so, we must have it disabled.

File: qiskit/transpiler/target.py [hunk 3]
```diff
@@ -350,8 +353,15 @@ def add_instruction(self, instruction, properties=None, name=None):
...
         else:
+            if None in properties:
+                self._global_operations[instruction.num_qubits].add(instruction_name)
             qargs_val = {}
             for qarg in properties:
+                if qarg is not None and len(qarg) != instruction.num_qubits:
+                    raise TranspilerError(
+                        f"The number of qubits for {instruction} does not match the number "
+                        f"of qubits in the properties dictionary: {qarg}"
+                    )
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:364]: This is probably out-of-scope for this PR, but some minor griping: the logic flow of this function is getting a little tricky for me now, since when I originally read this, I thought it would cause bugs with variadic instructions, as you can't do `type.num_qubits`.  In reality, it all works because `properties` is required to be `{None: None}`, so `qarg is not None` implies (transitively) that `instruction.num_qubits` should be valid, but because it's in a `for qarg in properties` _loop_, that's not so obvious to me any more.

Basically, I'm just griping that it's getting hard to keep all the different assumptions/implications about the objects in my head now.  There's nothing actionable unless there's an obvious way to refactor the logic a little now that we've been forced to support variadic arguments.

**Review by jakelishman — APPROVED** [review]
On the whole, this looks good to merge to me, and I'm glad it also handles variadic instructions correctly.  My only comments are minor questions / gripes that don't actually need any action within this PR.

**Review by ajavadia — APPROVED** [review]
LGTM

File: qiskit/transpiler/target.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:557]: I don't see it in the global disable list: https://github.com/Qiskit/qiskit-terra/blob/main/.pylintrc#L61-L76 and the only file specific disable is `too-many-return-statements`. I can change this to `[]` if you'd prefer. I think I just used list because the other method used `set()` and it was in my head.

File: qiskit/transpiler/target.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:364]: Yeah, I feel like I need to write a big comment about all the different internal states and ways we track things. Part of it just is there are a lot of fiddly details here as there are basically 4 classes of instructions: normal instructions that have qubit restrictions either with or without properties, globally defined ideal instances of operations that have qubit and parameter restrictions, and globally defined variable width instructions. We've basically grafted the last two of those onto the target model which was designed for the first two (the 3rd class was originally just a shorthand to support ideal simulation without adding every permutation of qubit with no properties payload).

In this particular case I had tried to do this check outside the loop but I had logic flow issues with that as we still need to run this loop for the global fixed width instruction case even if this check itself isn't valid for those cases. I think we can look at trying to improve this in a separate refactor in a followup, especially as that I wouldn't want to backport necessarily.

File: qiskit/transpiler/target.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:243]: In general we haven't been putting any constraints in the target around classical bits. Everything in the target so far has been quantum cosntraints which is why I just used the number of qubits as the dict key here.

File: qiskit/transpiler/target.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/transpiler/target.py:557]: No, I don't care - I just thought pylint had moaned at me before. Maybe it's in a newer version?

File: qiskit/transpiler/target.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/transpiler/target.py:243]: Sounds good!

File: qiskit/transpiler/target.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/transpiler/target.py:364]: Yeah, it's absolutely a follow-up thing, if we do it at all.
